util/pyclient/bft_client.py

```diff
@@ -241,3 +247,256 @@ def get_rsi_replies(self):
+    authenticity and integrity. To enable a TLS-handshake, it uses self-signed SSL X.509 certificates.
+    """
+    # In create_tls_certs.sh - openssl command line utility uses CN(certificate name) in the subj field.
+    # This is the host name (domain name) to be verified.
+    CERT_DOMAIN_FORMAT="node%dser"
+    # Taken from TlsTCPCommunication.cpp (we prefer hard-code and not to parse the file)
+    MSG_HEADER_SIZE=4
+
+    def __init__(self, config, replicas):
+        super().__init__(config, replicas)
+        self.ssl_streams = dict()
+
+    async def on_started_replicas(self, replica_ids):
+        """
+        Establish a TLS over TCP connections to all replicas concurrently (using a nursery). 
+        We relay on caller cancel scope.
+        """
+        async with trio.open_nursery() as nursery:
+            for id in replica_ids:
+                replica = self.replicas[id]
+                if (replica.ip, replica.port) not in self.ssl_streams.keys():
+                    nursery.start_soon(self._establish_ssl_stream, replica)
+
+    def _get_private_key_path(self, replica_id, is_client=False):
+        """
+        Private key is under <certificate root path>/replica_id/<node type>/pk.pem,
+        where node type is "server" or "client".
+        """
+        cert_type = "client" if is_client else "server"
+        return os.path.join(self.config.certs_path, str(replica_id), cert_type, "pk.pem")
+
+    def _get_cert_path(self, replica_id, is_client=False):
+        """
+        Certificate is under <certificate root path>/replica_id/<node type>/cert.pem,
+        where node type is "server" or "client".
+        """
+        cert_type = "client" if is_client else "server"
+        return os.path.join(self.config.certs_path, str(replica_id), cert_type, cert_type + ".cert")
+
+    async def _establish_ssl_stream(self, dest_replica):
+        """
+        Try every 250ms to connect to dest_replica. We relay on the caller context that will trigger cancalleation at 
+        some point in time. Break the loop on success and keep the ssl_stream in self.ssl_streams for future use.
+        """
+        server_cert_path = self._get_cert_path(dest_replica.id, False)
+        client_cert_path = self._get_cert_path(self.client_id, True)
+        client_pk_path = self._get_private_key_path(self.client_id, True)
+        # Create an SSl context - enable CERT_REQUIRED and check_hostname = True
+        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        # Verify server certificate using this trusted path
+        ssl_context.load_verify_locations(cafile=server_cert_path)
+        # Load my private key and certificate
+        ssl_context.load_cert_chain(client_cert_path, client_pk_path)
+        # Server hostname to be verified must be taken from create_tls_certs.sh
+        server_hostname = self.CERT_DOMAIN_FORMAT % dest_replica.id
+        while True:
+            try:
```

Comment:
to avoid an infinite wait, does it make sense to wrap with:
```python
with trio.fail_after()
```
?

Reply:
There is a caller fail_after block in bft.py . I wrote a comment in on_started_replicas().
I'm still not sure about the right behavior. 
In the trio guide, the say a good implementation should not put fail_after for reusable functions.
My idea is that a good one should give an optional fail after timeput. But I'm not sure if it's possible to nicely implement a conditional fail_after (can also put infinite time if the user doesn't want to fail, but it's ugly ).

Reply:
I think leaving it to the caller is the right move, as they say in the trio docs. Let the caller decide how long to wait. That caller can also decide to scope this call inside an outer nursery etc.. 

When I originally wrote the UDP code I added some timeouts, but I think following the trio guidelines makes the most sense. The author has spent years working on this stuff and thinking about the proper abstractions.

Reply:
@andrewjstone I think I agree with you. But there are some problems with each methodology:
Let's say you put your own timeout: The big issue is - if you don't look at the whole code including the caller's code, you can't even know if your own timeout is effective. I've seen some examples where the caller put 5-sec delay, and the inside code puts a 10-sec delay. This can happen if you decide to change your code after the caller wrote its own code.
On the other hand, letting the caller decide might be an issue as well, in complicated code- many times caller might not have any good idea about "what is a good timeout". So the caller must have a very good idea on what's going on inside.